src/err/mod.rs

```diff
@@ -844,12 +833,31 @@ mod tests {
     }
 
     #[test]
+    #[cfg(Py_3_12)]
+    fn invalid_error_type() {
+        Python::with_gil(|py| {
+            let err: PyErr = PyErr::new::<crate::types::PyString, _>(());
+            assert!(err.is_instance_of::<exceptions::PySystemError>(py));
+            err.restore(py);
+            let err = PyErr::fetch(py);
+
+            assert!(err.is_instance_of::<exceptions::PySystemError>(py));
+            assert_eq!(
+                err.to_string(),
+                "SystemError: _PyErr_SetObject: exception <class 'str'> is not a BaseException subclass"
```

Comment:
While this is a different error message to the non-3.12 case, I think it's actually a better error message. If the difference is undesirable I can either:
- Update the 3.12 case to check and return our exception, or
- Make normalization on non-3.12 cases _also_ use `PyErr_SetObject`, which would change the exception message on all versions. (Maybe I should benchmark this, and consider splitting out into a separate PR as it would be breaking.)

For reference, the assertion in the non-3.12 equivalent is:

```rust
            assert_eq!(
                 err.to_string(),
                "TypeError: exceptions must derive from BaseException"
            );
```

Reply:
For now I've just kept it the same in 3.12 as before, we could reconsider changing this another time.